invenio_records_rest/serializers/csv.py

```diff
@@ -0,0 +1,118 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2016-2018 CERN.
+#
+# Invenio is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Marshmallow based DublinCore serializer for records."""
+
+from __future__ import absolute_import, print_function
+
+from .base import PreprocessorMixin, SerializerMixinInterface
+from .marshmallow import MarshmallowMixin
+
+
+class CSVSerializer(SerializerMixinInterface, MarshmallowMixin,
+                    PreprocessorMixin):
+    """CSV serializer for records.
+
+    Note: This serializer is not suitable for serializing large number of
+    records.
+    """
+
+    def serialize(self, pid, record, links_factory=None, filter_keys=None):
+        """Serialize a single record and persistent identifier.
+
+        :param pid: Persistent identifier instance.
+        :param record: Record instance.
+        :param links_factory: Factory function for record links.
+        :param filter_keys: Dictionary of keys to filter out
+                            with optional values to keep
+        """
+        record = self.process_dict(
+            self.transform_record(pid, record, links_factory),
+            filter_keys=filter_keys)
+
+        return self.format_csv([record])
+
+    def serialize_search(self, pid_fetcher, search_result, links=None,
+                         item_links_factory=None, filter_keys=None):
+        """Serialize a search result.
+
+        :param pid_fetcher: Persistent identifier fetcher.
+        :param search_result: Elasticsearch search result.
+        :param links: Dictionary of links to add to response.
+        :param item_links_factory: Factory function for record links.
+        :param filter_keys: key = dictionary of keys to filter out
+                            value = optional field to keep from the dictionary
+                            of values associated with the key
+        """
+        records = []
+        for hit in search_result['hits']['hits']:
+            processed_hit = self.transform_search_hit(
+                pid_fetcher(hit['_id'], hit['_source']),
+                hit,
+                links_factory=item_links_factory,
+            )
+            records.append(self.process_dict(processed_hit, filter_keys))
+
+        return self.format_csv(records)
+
+    def format_csv(self, records):
+        """Format list of flattened dictionaries into CSV format."""
+        headers = sorted(set([key for record in records for key in record]))
+
+        rows = [["\"" + str(record[key]) + "\"" if key in record else "" for key in headers] for
```

Comment:
I think we can try to optimize this, maybe we can quickly discuss this together. IMHO you could try to build your `set` of keys for the headers while iterating and building the rows.